crates/oauth2-types/src/oidc.rs

```diff
@@ -192,44 +287,588 @@ pub struct Metadata {
+    #[error("supported response types are missing")]
+    MissingResponseTypesSupported,
+
+    /// The supported subject types are missing.
+    #[error("supported subject types are missing")]
+    MissingSubjectTypesSupported,
+
+    /// The supported ID token signing algorithm values are missing.
+    #[error("supported ID token signing algorithm values are missing")]
+    MissingIdTokenSigningAlgValuesSupported,
+
+    /// The URL of the given field doesn't use a `https` scheme.
+    #[error("{0}'s URL doesn't use a https scheme: {1}")]
+    UrlNonHttpsScheme(&'static str, Url),
+
+    /// The URL of the given field contains a query, but it's not allowed.
+    #[error("{0}'s URL contains a query: {1}")]
+    UrlWithQuery(&'static str, Url),
+
+    /// The URL of the given field contains a fragment, but it's not allowed.
+    #[error("{0}'s URL contains a fragment: {1}")]
+    UrlWithFragment(&'static str, Url),
+
+    /// The issuer URL doesn't match the one that was discovered.
+    #[error("issuer URLs don't match")]
+    IssuerUrlsDontMatch,
+
+    /// `openid` is missing from the supported scopes.
+    #[error("missing openid scope")]
+    ScopesMissingOpenid,
+
+    /// `code` is missing from the supported response types.
+    #[error("missing `code` response type")]
+    ResponseTypesMissingCode,
+
+    /// `id_token` is missing from the supported response types.
+    #[error("missing `id_token` response type")]
+    ResponseTypesMissingIdToken,
+
+    /// `id_token token` is missing from the supported response types.
+    #[error("missing `id_token token` response type")]
+    ResponseTypesMissingIdTokenToken,
+
+    /// `authorization_code` is missing from the supported grant types.
+    #[error("missing `authorization_code` grant type")]
+    GrantTypesMissingAuthorizationCode,
+
+    /// `implicit` is missing from the supported grant types.
+    #[error("missing `implicit` grant type")]
+    GrantTypesMissingImplicit,
+
+    /// The given endpoint is missing auth signing algorithm values, but they
+    /// are required because it supports at least one of the `client_secret_jwt`
+    /// or `private_key_jwt` authentication methods.
+    #[error("{0} auth signing algorithm values contain `none`")]
+    MissingAuthSigningAlgValues(&'static str),
+
+    /// `none` is in the given endpoint's signing algorithm values, but is not
+    /// allowed.
+    #[error("{0} signing algorithm values contain `none`")]
+    SigningAlgValuesWithNone(&'static str),
+
+    /// `RS256` is missing from the given endpoint's signing algorithm values.
+    #[error("missing RS256 in {0} signing algorithm values")]
+    SigningAlgValuesMissingRs256(&'static str),
+}
+
+/// Possible extra restrictions on a URL.
+#[derive(Debug, PartialEq, Eq)]
+enum ExtraUrlRestrictions {
+    /// The URL must not contain a fragment.
+    NoFragment,
+
+    /// The URL must not contain a query or a fragment.
+    NoQueryOrFragment,
+}
```

Comment:
What about including a `None` possible value here, to avoid the option? It would make the `validate_url` logic simpler

Something like

```rust
/// Possible extra restrictions on a URL.
#[derive(Debug, PartialEq, Eq, Copy)]
enum UrlRestrictions {
    None,

    /// The URL must not contain a fragment.
    NoFragment,

    /// The URL must not contain a query or a fragment.
    NoQueryOrFragment,
}

impl UrlRestrictions {
    const fn can_have_fragment(self) -> bool {
       self == Self::None
    }

    const fn can_have_query(self) -> bool {
       self != Self::NoQueryFragment
    }
}

fn validate_url(
     field: &'static str,
     url: &Url,
     restrictions: UrlRestrictions,
 ) -> Result<(), ProviderMetadataVerificationError> {
     if url.scheme() != "https" {
         return Err(ProviderMetadataVerificationError::UrlNonHttpsScheme(
             field,
             url.clone(),
         ));
     }

 
     if !restrictions.can_have_query() && url.query().is_some() {
         return Err(ProviderMetadataVerificationError::UrlWithQuery(
             field,
             url.clone(),
         ));
     }

     if !restrictions.can_have_fragment() && url.fragment().is_some()
     {
         return Err(ProviderMetadataVerificationError::UrlWithFragment(
             field,
             url.clone(),
         ));
     }

     Ok(())
 }

```